cmd/k8s-operator/exec.go

```diff
@@ -0,0 +1,929 @@
+	resizeStreamID atomic.Uint32
+
+	wmu    sync.Mutex // sequences writes
+	closed bool
+
+	rmu                 sync.Mutex // sequences reads
+	writeCastHeaderOnce sync.Once
+
+	zlibReqReader zlibReader
+	// writeBuf is used to store data written to the connection that has not
+	// yet been parsed as SPDY frames.
+	writeBuf bytes.Buffer
+	// readBuf is used to store data read from the connection that has not
+	// yet been parsed as SPDY frames.
+	readBuf bytes.Buffer
+	log     *zap.SugaredLogger
+}
+
+// Reads parses bytes passed to it as SPDY frames and, frame by frame, passes
+// them to the original destination. If the frame is a SPDY data frame for
+// resize stream, also sends its payload to the session recorder.
+func (c *spdyRemoteConnRecorder) Read(b []byte) (int, error) {
+	log := c.log.With("spdyRemoteConnRecorder", "read")
+	c.rmu.Lock()
+	defer c.rmu.Unlock()
+	n, err := c.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+	c.readBuf.Write(b[:n])
+
+	var sf spdyFrame
+	ok, err := sf.Parse(c.readBuf.Bytes(), log)
+	if err != nil {
+		return 0, fmt.Errorf("error parsing data read from connection: %w", err)
+	}
+	if !ok {
+		// The parsed data in the buffer will be processed together with
+		// the new data on the next call to Read.
+		return n, nil
+	}
+	c.readBuf.Next(len(sf.Raw)) // advance buffer past the parsed frame
+
+	if !sf.Ctrl { // data frame
+		switch sf.StreamID {
+		case c.resizeStreamID.Load():
+			var err error
+			c.writeCastHeaderOnce.Do(func() {
+				var resizeMsg struct {
+					Width  int `json:"width"`
+					Height int `json:"height"`
+				}
+				if err = json.Unmarshal(sf.Payload, &resizeMsg); err != nil {
+					return
+				}
+				c.ch.Width = resizeMsg.Width
+				c.ch.Height = resizeMsg.Height
+				var j []byte
+				j, err = json.Marshal(c.ch)
+				if err != nil {
+					return
+				}
+				j = append(j, '\n')
+				_, err = c.lw.sessionRecorder.Write(j)
+				if err != nil {
+					c.log.Debugf("received error from recorder: %v", err)
+				}
+			})
+			if err != nil {
+				return 0, err
+			}
+		}
+		return n, nil
+	}
+	// We always want to parse the headers, even if we don't care about the
+	// frame, as we need to advance the zlib reader otherwise we will get
+	// garbage.
+	header, err := sf.parseHeaders(&c.zlibReqReader, log)
+	if err != nil {
+		return 0, fmt.Errorf("error parsing frame headers: %w", err)
+	}
+	if sf.Type == SYN_STREAM {
+		c.storeStreamID(sf, header)
+	}
+	return n, nil
+}
+
+// Write forwards the raw data of the latest parsed SPDY frame to the original
+// destination. If the frame is an SPDY data frame, it also sends the payload to
+// the connected session recorder.
+func (c *spdyRemoteConnRecorder) Write(b []byte) (int, error) {
+	log := c.log.With("spdyRemoteConnRecorder", "write")
+	c.wmu.Lock()
+	defer c.wmu.Unlock()
+	c.writeBuf.Write(b)
+
+	var sf spdyFrame
+	ok, err := sf.Parse(c.writeBuf.Bytes(), log)
+	if err != nil {
+		return 0, fmt.Errorf("error parsing data: %w", err)
```

Comment:
Here we could also, if the policy was to fail open, forward the raw bytes anyway (and stop attempting to record) the next reads and writes. Perhaps we can revisit that part in the next iteration